### PR TITLE
Category page

### LINKED ISSUE #23: Category Page
## Setup Model 

* [x] model and fields: 
  * name:string, visible:boolean
  * a category has many products

### Product model
* [x] add relationship between product and category

### Setup Mirage
* [x] update and prepare factories 
* [x] update default scenario (it's for development) 

## Listing
* [x] routes & template - `categories` this.store.findAll('category')
* [ ] component `category-thumb` (basic info; link to category products listing) (ensure it renders integration test)

## Show
* [x] routes & template - `categories.category` /category/1/ this.store.find('category', id) 
* [x] component `category-show` (products of given category) (ensure it renders integration test)

## Category form component
* [x] [changeset](https://github.com/DockYard/ember-changeset) & [validation](https://github.com/DockYard/ember-changeset-validations/) 
* [x] action that save changes 

## Edit
* [x] routes & template `categories.category.edit` /category/1/edit
* [x] `category-form` component 

## Add 
* [x] routes & template `categories.new` /category/new
* [x] `category-form` component 

## Delete
* [x] action on `categories` ( destroy|delete & save )
* category can only be deleted if it has no products
* [x] ask for confirmation 

## Acceptance Tests 
* [x] listing works fine 
* [x] you add new
* [x] you edit 
* [x] you can delete

### PR DESCRIPTION
**Pull request by MattSkrobis**
Closes #23

### CODE REVIEW

File: TIL.md [hunk 1]
```diff
@@ -1,3 +1,12 @@
+# Thursday
```
Comment (cs3b) [inline TIL.md]: What about Friday?

File: app/components/category-destroy/component.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import Ember from 'ember';
+
+const { Component, inject: { service } } = Ember;
+
+export default Component.extend({
+  router: service(),
+  actions: {
+    destroy(category) {
+      category.deleteRecord();
+      category.save();
```
Comment (cs3b) [inline app/components/category-destroy/component.js]: You should do this as a callback, so only after you get confirmation from the backend. 
Together with that it's good to handle case when delete is unsuccessfull

File: app/components/category-form/component.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import Ember from 'ember';
+import { categoryValidations } from 'bsc-ember/validations/category';
+import Changeset from 'ember-changeset';
+import lookupValidator from 'ember-changeset-validations';
+
+const { Component,
```
Comment (cs3b) [inline app/components/category-form/component.js]: keep with one style, or use prettier plugin for visual studio code (I've tested it during the weekend and works nicely)

File: app/components/category-form/component.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  inject: {
+    service
+  } } = Ember;
+
+export default Component.extend({
+  router: service(),
+  init() {
+    this._super(...arguments);
+    this.changeset = new Changeset(this.model, lookupValidator(categoryValidations), categoryValidations);
```
Comment (cs3b) [inline app/components/category-form/component.js]: it's might be better if we will do this on `didReceiveAttrs`

Do you know why?

File: app/components/category-show/template.hbs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{{yield}}
+<h1 data-test-category-show-name> {{model.name}} </h1>  
+
+<div> Products in this category </div>
+<ul  data-test-category-show-products>
```
Comment (cs3b) [inline app/components/category-show/template.hbs]: double space ... do template linter works ?

File: app/controllers/categories/index.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+import Ember from 'ember';
+
+const { Controller, inject: { service } } = Ember;
+
+export default Controller.extend({
+  router: service(),
+  actions: {
+    toNew() {
```
Comment (cs3b) [inline app/controllers/categories/index.js]: why do you need this action?

File: mirage/factories/product.js [hunk 7]
```diff
@@ -1,4 +1,4 @@
-import { Factory, faker } from 'ember-cli-mirage';
+import { Factory, faker  } from 'ember-cli-mirage';
```
Comment (cs3b) [inline mirage/factories/product.js]: double space

File: tests/acceptance/categories-test.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    application = startApp();
+  });
+
+  afterEach(function() {
+    destroyApp(application);
+  });
+
+  it('can visit /categories', async function() {
+    server.create('category', { name: 'Cheap' });
+    await page.visit();
```
Comment (cs3b) [inline tests/acceptance/categories-test.js]: it's not bad to ensure that data is visible on the listing

File: tests/pages/category.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      open: clickable('[data-test-category-show-link]'),
+      thumbName: text('[data-test-category-thumb-name]')
+    }
+  }),
+  visit: visitable('/categories'),
+  categoryName: text('[data-test-category-show-name]'),
+  openNewCategory: clickable('[data-test-categories-new-link]'),
+  openCategories: clickable('[data-test-categories-link]'),
+  edit: clickable('[data-test-category-edit-link]'),
+  editCategoryName: fillable('input', { at: 0 }),
```
Comment (cs3b) [inline tests/pages/category.js]: consider `editCategoryName` vs `categoryNameInput`

File: app/controllers/categories/index.js (hunk 6 above)
Comment (MattSkrobis) [inline app/controllers/categories/index.js]: I wanted to use paper-button and I failed to see a neater way to make it a 'link' than triggering an action on click

File: tests/acceptance/products-test.js [hunk 10]
```diff
@@ -20,47 +20,40 @@ describe('Acceptance | products', function() {
...
-    // show 
+    // show
     await page.products(0).open();
     expect(page.productName).to.include('Personal Computer');
 
     // edit
-    await page.openProducts().edit()
-      .editProductName('Apple 1')
-      .saveProduct();
+    await page.openProducts().edit().editProductName('Apple 1').saveProduct();
```
Comment (cs3b) [inline tests/acceptance/products-test.js]: would be nice to change prettier in the way that each `.functionCall` would be in new line ...

**Pull request merged**